### PR TITLE
Allow relative z in multi_d_acquisition_events

### PR DESCRIPTION
**Pull request by ianhi**
Closes: https://github.com/micro-manager/pycro-manager/issues/153

This should maintain all the previous behaviors but also allow for specifying z positions in these new ways:

- z_positions as 2D array
- z_positions as 1D array
- z_positions as 1D array + relative positions

which is hopefully made clear in the new docs page I added for this. 

Stuff like this is a great candidate for pytest, I thought about adding that but didn't becuase you already have a large user automated testing DIR that I wasn't sure would work with pytest on CI. Let me know what would be best

### CODE REVIEW

**Comment by henrypinkard** [comment]
A couple things: 
 - The contents and the description of the notebook are not in the same order on the applications page
 - In the notebook I think you can delete the part with `np.c_` and just show `hstack`, since this is showing a feature of numpy, not pycromanager
 - I was a bit confused about the docstring for `multi_d_acquisition_events`, specifically the  "for each xy point, or 2D (shape: (N, n_z) ) to fully specify the xyz points". After reading the notebook I get what `n_z` is supposed to be. But I think it might be a simpler interface for this to have `xyz_positions` and `xy_positions`, where the former automatically puts you into relative z mode and any z stacks are relative to it, while the latter defaults to absolute z mode if a z stack is provided. This wway i think would fit in more with the existing API, and it would leave less work for the user in the case of different z stacks at different positions (i.e. just supply `xyz_positions` and `z_start`, `z_end`, `z_step` without having to construct the 2d array yourself). What do you think?
 - The additional descriptions of channels/time in the notebook is really nice. It could be good to link to this notebook from the [features page on MDAs](https://pycro-manager--158.org.readthedocs.build/en/158/acq_events.html#multi-dimensional-acquisitions)

**Comment by henrypinkard** [comment]
Also, yes, it would be great to some more stuff built into automatic testing. That test directory is not automated at the moment, I just copy and paste them to quickly debug when I think things are broken. @bryantChhun has also been thinking about adding automated testing. Maybe better to continue this conversation [on the issue for it](https://github.com/micro-manager/pycro-manager/issues/26)

**Comment by henrypinkard** [comment]
> > The additional descriptions of channels/time in the notebook is really nice. It could be good to link to this notebook from the features page on MDAs
> 
> I'm not sure I understand what the order is meant to be. Is it chronological? Also, if anything this notebook is more of a tutorial than an application, would it make sense to fold it into the page on MDAs?

Vaguely scales with complexity. Either position in the ordering is fine, you've just got it first in the TOC but second on the actual list right now (or vice versa). I think given the density of it, it makes more sense to have it with applications. 

> > I was a bit confused about the docstring for multi_d_acquisition_events, specifically the "for each xy point, or 2D (shape: (N, n_z) ) to fully specify the xyz points". After reading the notebook I get what n_z is supposed to be. But I think it might be a simpler interface for this to have xyz_positions and xy_positions, where the former automatically puts you into relative z mode and any z stacks are relative to it, while the latter defaults to absolute z mode if a z stack is provided. This wway i think would fit in more with the existing API, and it would leave less work for the user in the case of different z stacks at different positions (i.e. just supply xyz_positions and z_start, z_end, z_step without having to construct the 2d array yourself). What do you think?
> 
> I was probably doing some premature optimization here, the current setup allows for different non-uniform stepped z stacks for each positions. So you could have
> pos1: z = [-3,-1,0]
> pos2: z= [-10, -1, 0]
> 
> but that's probably not necessary to include.

Makes sense. Generally speaking I don't think its worth the effort to try to get every edge case into this function, since for customized stuff you should probably just be going a level deeper and creating events manually anyway. In my mind z-stacks with custom nonuniform step spacing would fall under this

> To be explicit you're proposing the following call signatures:
> 
> ```python
> # Gives relative z
> multi_d_acquisition_events(xyz_positions = xyz)
> multi_d_acquisition_events(xyz_positions = xyz, z_start=-1, z_stop = 1, z_step=1)
> 
> # absolute z
> multi_d_acquisition_events(z_start=0, z_stop=10,z_step=1)
> multi_d_acquisition_events(xy_positions = xy, z_start=0, z_stop=10,z_step=1)
> ```

Yeah exactly. Or maybe a `relative_z=True` argument could be added to make this explicit. What do you prefer?

> I'd suggest adding the option of an array of relative z values:
> (this should give the same result as xyz_positions passed with z_start et al)
> 
> ```python
> multi_d_acquisition_events(xyz_positions = xyz, z_rel = np.linspace(-1,1,3))
> ```

I think its better to make an example of doing this manually than to have an explicit option. The purpose of this function is a a convenience for people who dont want to dig any deeper into the API, so I think the fewer ways of doing things, the better

> The only downsides I see are:
> 
>     1. you need to know current xy point in order to make a single position z stack

Maybe this is an argument for `relative_z=True` argument

>     2. not totally clear to me how best to handle an order string like `zp`
>        

Fine to just throw in an exception in the cases that it doesnt make sense

**Comment by ianhi** [comment]
@henrypinkard I think this is good now. I ended up adding github actions to run the pytest tests I added for MDA, but I think you need to change a setting to get them to run on PRs. I opened a PR against my fork and the tests all ran there: https://github.com/ianhi/pycro-manager/pull/1 I think you need to modify these settings: 
image

**Comment by ianhi** [comment]
> Yeah exactly. Or maybe a relative_z=True argument could be added to make this explicit. What do you prefer?

I went for raising errors rather than adding yet more kwargs.


> Makes sense. Generally speaking I don't think its worth the effort to try to get every edge case into this function,

sounds good! It indeed simplified it greatly.


>  @bryantChhun has also been thinking about adding automated testing. Maybe better to continue this conversation on the issue for it

In a moment of forgiveness is better than permission I went ahead and greated github actions + pytest tests for the content of this PR. Testing the actual functionality of the integration with μmanager would be much more complex though so I'll leave that to the other issue. 

pytest supports multiple test directories and I think its pretty standard to have one inside the `pycromanager` folder so I added one there for testing things that are contained to python, should be easy enough to add in the other test directory or combine them in future. But I wanted to keep the automated and non-automated stuff separate for now.

File: pycromanager/acquire.py [hunk 1]
```diff
@@ -482,25 +504,36 @@ def multi_d_acquisition_events(num_time_points=1, time_interval_s=0, z_start=Non
...
+    if xy_positions is not None:
+        xy_positions = np.asarray(xy_positions)
+        z_positions = None
+    elif xyz_positions is not None:
+        xyz_positions = np.asarray(xyz_positions)
+        xy_positions = xyz_positions[:,:2]
+        z_positions = xyz_positions[:,2][:, None]
+
+    if has_zsteps:
+        z_rel = np.arange(z_start, z_end + z_step, z_step)
```
Comment (ianhi) [inline pycromanager/acquire.py:525]: This includes one more point than the old version. To me this way make more sense but it is clearly backwards incompatible.. let me know if you'd prefer this to be
```suggestion
        z_rel = np.arange(z_start, z_end, z_step)
```

**Comment by bryantChhun** [comment]
> In a moment of forgiveness is better than permission I went ahead and greated github actions + pytest tests for the content of this PR. Testing the actual functionality of the integration with μmanager would be much more complex though so I'll leave that to the other issue.

no worries!  I didn't do much other than add some mocks and a couple tests.  Will create a PR soon.  And in general, I never discourage anyone from writing tests for their code :-)

There was a discussion by email about the more complex testing with umanager (apologies, we agreed we'll move those discussions to github issues).  The general conclusion was that other umanager work [(release of 2.0)](https://github.com/micro-manager/micro-manager/milestone/1) is taking priority.  Continuous Integration with micromanager will require solving some issues in the [2.1 release.](https://github.com/micro-manager/micro-manager/milestone/4).

File: pycromanager/acquire.py (hunk 1 above)
Comment (henrypinkard) [reply pycromanager/acquire.py:525]: Yeah I think that's fine and probably an upgrade

**Comment by henrypinkard** [comment]
>  I think this is good now. I ended up adding github actions to run the pytest tests I added for MDA, but I think you need to change a setting to get them to run on PRs. I opened a PR against my fork and the tests all ran there: [ianhi#1](https://github.com/ianhi/pycro-manager/pull/1) I think you need to modify these settings:

I looked on the settings but it was the same as the picture you posted -- allow all actions was already selected.

**Comment by henrypinkard** [comment]
Any other ideas on how to set it up like you had in that link?

**Comment by ianhi** [comment]
> Any other ideas on how to set it up like you had in that link?

:thinking: hmm. I would have thought that was all that was necessary. I suppose there isn't an error in the yaml file as they ran on my fork. It looks as though they did run once you merged it:
image

I'm going to try opening a dummy PR to see if they will trigger on that now that they're properly in master, maybe its different than my repo bc I don't own it.

**Comment by ianhi** [comment]
See https://github.com/micro-manager/pycro-manager/pull/162. looks as though the tests work on newly created PR

**Pull request merged**